### PR TITLE
Deprecate `stochastic_weight_avg` from the `Trainer` constructor

### LINKED ISSUE #8971: Deprecate `stochastic_weight_avg` from Trainer constructor
## 🚀 Feature

Remove this argument from the Trainer constructor:
https://github.com/PyTorchLightning/pytorch-lightning/blob/c6b68883879e38719688865aceac746477f0a9b9/pytorch_lightning/trainer/trainer.py#L161

### Motivation

We are auditing the Lightning components and APIs to assess opportunities for improvements:
- https://docs.google.com/document/d/1xHU7-iQSpp9KJTjI3As2EM0mfNHHr37WZYpDpwLkivA/edit#
- https://github.com/PyTorchLightning/pytorch-lightning/issues/7740

First, from the user POV: why should users have to learn about 2 different ways of enabling stochastic weight averaging?
- Either add a trainer flag `stochastic_weight_avg=True`
- Or add a `StochasticWeightAveraging` callback to the `Trainer`'s `callbacks` list

This goes against the values here: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md#main-core-value-one-less-thing-to-remember

Secondly, this will simplify the internal codebase: https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md#simple-internal-code

The Trainer today has over 50 arguments to its constructor. This number is growing with each feature release, and makes the trainer cluttered. It also hurts the extensibility of the trainer: a number of arguments passed to the Trainer are to customize other utilities. Plumbing arguments through the trainer creates an undesirable coupling: when the underlying components change, the framework is forced to make breaking API changes in at least two places:
1. The underlying component
2. The trainer constructor

The trainer should have a high bar for what's included on the arguments list. `stochastic_weight_avg` configures a callback for users, which users could easily do in 1 line and pass into the `callbacks` argument.

By default, this is not enabled, so users have to opt into this anyways. 

### Pitch

1. Deprecate `stochastic_weight_avg ` off the Trainer constructor in v1.5
2. Update any tutorials/documentation to point to how users can easily construct a StochasticWeightAveraging callback on their own

https://github.com/PyTorchLightning/pytorch-lightning/blob/5329b0d11358b72a42d596d3b70b7010e35c45af/pytorch_lightning/trainer/connectors/callback_connector.py#L85-L93

3. Remove the flag in v1.7

This nets us 1 constructor argument saved, which is progress

### Additional context

______________________________________________________________________

#### If you enjoy Lightning, check out our other projects! ⚡

<sub>

- [**Metrics**](https://github.com/PyTorchLightning/metrics): Machine learning metrics for distributed, scalable PyTorch applications.

- [**Flash**](https://github.com/PyTorchLightning/lightning-flash): The fastest way to get a Lightning baseline! A collection of tasks for fast prototyping, baselining, finetuning and solving problems with deep learning

- [**Bolts**](https://github.com/PyTorchLightning/lightning-bolts): Pretrained SOTA Deep Learning models, callbacks and more for research and production with PyTorch Lightning and PyTorch

- [**Lightning Transformers**](https://github.com/PyTorchLightning/lightning-transformers): Flexible interface for high performance research using SOTA Transformers leveraging Pytorch Lightning, Transformers, and Hydra.

</sub>

### PR DESCRIPTION
**Pull request by ananthsub**
## What does this PR do?

Fixes #8971

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

### CODE REVIEW

File: CHANGELOG.md [hunk 1]
```diff
@@ -106,7 +106,7 @@ The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/).
 - Deprecated `DataModule` properties: `train_transforms`, `val_transforms`, `test_transforms`, `size`, `dims` ([#8851](https://github.com/PyTorchLightning/pytorch-lightning/pull/8851))
 
 
--
+- Deprecated passing `stochastic_weight_avg` from the `Trainer` constructor ([#8989](https://github.com/PyTorchLightning/pytorch-lightning/pull/8989))
```
Comment (awaelchli) [inline CHANGELOG.md:109]:
```suggestion
- Deprecated passing `stochastic_weight_avg` from the `Trainer` constructor in favor of adding the `StochasticWeightAveraging` callback directly to the list of callbacks ([#8989](https://github.com/PyTorchLightning/pytorch-lightning/pull/8989))
```

File: pytorch_lightning/trainer/connectors/callback_connector.py [hunk 2]
```diff
@@ -39,6 +40,12 @@ def on_trainer_init(
         # init folder paths for checkpoint + weights save callbacks
         self.trainer._default_root_dir = default_root_dir or os.getcwd()
         self.trainer._weights_save_path = weights_save_path or self.trainer._default_root_dir
+        if stochastic_weight_avg:
+            rank_zero_deprecation(
+                "Setting `Trainer(stochastic_weight_avg=True)` is deprecated in v1.5 and will be removed in v1.7."
+                " Please pass `pytorch_lightning.callbacks.stochastic_weight_avg.StochasticWeightAveraging`"
+                " directly to the Trainer's ``callbacks`` argument instead."
```
Comment (awaelchli) [inline pytorch_lightning/trainer/connectors/callback_connector.py:47]:
```suggestion
                "Setting `Trainer(stochastic_weight_avg=True)` is deprecated in v1.5 and will be removed in v1.7."
                " Please pass `pytorch_lightning.callbacks.stochastic_weight_avg.StochasticWeightAveraging`"
                " directly to the Trainer's `callbacks` argument instead."
```
for the python message

**Review by awaelchli — APPROVED** [review]

**Review by justusschock — APPROVED** [review]

**Review by tchaton — COMMENTED** [review]
@williamFalcon @edenlightning  Do you validate this depreciation ?

**Comment by ananthsub** [comment]
> @williamFalcon @edenlightning Do you validate this depreciation ?

@tchaton isn't this what the leads group is for? I'd appreciate input from @Borda @ethanwharris @SeanNaren @carmocca @kaushikb11 either here or on https://github.com/PyTorchLightning/pytorch-lightning/issues/8971 where I've listed the motivation & pitch

**Review by SkafteNicki — APPROVED** [review]

**Comment by carmocca** [comment]
Any deprecations should also remove the associated docs for the deprecated feature. Following/Copying a docs snippet to find out it's deprecated is bad UX.

For this case, the following would need to be removed:

https://github.com/PyTorchLightning/pytorch-lightning/blame/master/docs/source/advanced/training_tricks.rst#L54-L74

Part of the text there could be merged with the docstring for the callback

**Comment by carmocca** [comment]
> isn't this what the leads group is for?

The leads are meant to push the project's direction and explore new areas of improvement (in addition to the basic contributing work).

To green-light all these deprecations, we need a final decision by our beloved [BDFL](https://pytorch-lightning.readthedocs.io/en/latest/governance.html#bdfl). Let's continue this discussion in #9006

**Comment by tchaton** [comment]
> > @williamFalcon @edenlightning Do you validate this depreciation ?
> 
> @tchaton isn't this what the leads group is for? I'd appreciate input from @Borda @ethanwharris @SeanNaren @carmocca @kaushikb11 either here or on #8971 where I've listed the motivation & pitch

This decision was a strong push and final decision from Will. More importantly, I believe @edenlightning which represents the entire community should be present for strong Trainer depreciation.

Personally, I do agree with this change.

Best,
T.C

**Review by Borda — APPROVED** [review]

**Comment by Borda** [comment]
@ananthsub mind resolve conflicts? :]

**Review by daniellepintz — APPROVED** [review]
Looks great! @ananthsub can we resolve conflicts and merge?
